lahja/endpoint.py

```diff
@@ -33,62 +38,101 @@
         self._queues: Dict[Type[BaseEvent], List[asyncio.Queue]] = {}
-        self._executor: Optional[ThreadPoolExecutor] = None
         self._running = False
-        self._optional_internal_queue: Optional[asyncio.Queue] = None
-        self._optional_loop: Optional[asyncio.AbstractEventLoop] = None
-        self._optional_receiving_loop_running: Optional[asyncio.Event] = None
-        self._optional_internal_loop_running: Optional[asyncio.Event] = None
 
     @property
-    def _loop(self) -> asyncio.AbstractEventLoop:
-        if self._optional_loop is None:
-            raise NoConnection("Need to connect first")
-        return self._optional_loop
+    def ipc_path(self) -> pathlib.Path:
+        return self._ipc_path
 
     @property
-    def _internal_queue(self) -> asyncio.Queue:
-        if self._optional_internal_queue is None:
-            raise NoConnection("Need to connect first")
-        return self._optional_internal_queue
+    def name(self) -> str:
+        return self._name
 
-    @property
-    def _receiving_loop_running(self) -> asyncio.Event:
-        if self._optional_receiving_loop_running is None:
-            raise NoConnection("Need to connect first")
-        return self._optional_receiving_loop_running
-
-    @property
-    def _internal_loop_running(self) -> asyncio.Event:
-        if self._optional_internal_loop_running is None:
-            raise NoConnection("Need to connect first")
-        return self._optional_internal_loop_running
-
-    def connect_no_wait(self, loop: Optional[asyncio.AbstractEventLoop] = None) -> None:
+    def connect_no_wait(self,
+                        connection_config: ConnectionConfig,
+                        loop: Optional[asyncio.AbstractEventLoop] = None) -> None:
         """
-        Connect the :class:`~lahja.endpoint.Endpoint` to the :class:`~lahja.eventbus.EventBus`
-        instance that created this endpoint.
+        Connect the :class:`~lahja.endpoint.Endpoint` so that it can receive events.
         """
         if loop is None:
             loop = asyncio.get_event_loop()
 
-        self._optional_loop = loop
-        self._optional_internal_loop_running = asyncio.Event(loop=self._loop)
-        self._optional_receiving_loop_running = asyncio.Event(loop=self._loop)
-        self._optional_internal_queue = asyncio.Queue()
+        self._name = connection_config.name
+        self._ipc_path = connection_config.path
+        self._create_external_api(self._ipc_path)
+        self._loop = loop
+        self._internal_loop_running = asyncio.Event(loop=self._loop)
+        self._receiving_loop_running = asyncio.Event(loop=self._loop)
+        self._internal_queue = asyncio.Queue(loop=self._loop)
+        self._receiving_queue = asyncio.Queue(loop=self._loop)
 
         # Using `gather` (over e.g. `wait` or plain `ensure_future`) ensures that the inner futures
         # are automatically cancelled as soon as the parent task is cancelled
```

Comment:
Not part of this PR but the lines below this confuse me:

```python
asyncio.gather(
    asyncio.ensure_future(self._connect_receiving_queue(), loop=self._loop),
    asyncio.ensure_future(self._connect_internal_queue(), loop=self._loop),
    loop=self._loop
)
```

Doesn't this cause a warning? I think that `ensure_future` schedules these tasks to be run, but asyncio.gather() creates a task which is never awaited.

Reply:
I remember I had put it there to get rid of a cancel that I was seeing in Trinity when I did not have it. The point about `gather` is that it ensures all child tasks are cancelled when the gather task is cancelled.

From the docs: https://docs.python.org/3/library/asyncio-task.html#asyncio.gather
>If gather() is cancelled, all submitted awaitables (that have not completed yet) are also cancelled.

But I see what you mean, it looks weird. Definitely worth taking another look!

Reply:
Oh, interesting. I don't think the gather will ever be cancelled, since it's never scheduled, but maybe this binds together the fates of these two coros? If `_connect_receiving_queue` is canceled then `_connect_internal_queue` will be as well?

Thanks for the pointer, I'll play with it sometime